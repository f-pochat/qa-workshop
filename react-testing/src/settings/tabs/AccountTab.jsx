import React from 'react'
import { Button } from "@/components/ui/button"
import {
  Card,
  CardContent,
  CardDescription,
  CardHeader,
  CardTitle,
} from "@/components/ui/card"
import {
  Form,
  FormControl,
  FormDescription,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "@/components/ui/form"
import { Input } from "@/components/ui/input"
import { zodResolver } from "@hookform/resolvers/zod"
import { useForm } from "react-hook-form"
import { z } from "zod"

const formSchema = z.object({
	name: z.string().min(2, "Name too short").max(50, "Name too long"),
	username: z.string().min(2, "Username too short").max(50, "Username too long"),
})

export const AccountTab = () => {
	const form = useForm({
		resolver: zodResolver(formSchema),
		defaultValues: {
			name: 'Cianca',
			username: 'dinos1337'
		}
	})

	const onSubmit = (values) => {
		// some kind of API call should happen here.
		console.log({ values })
	}

	return (
		<Card>
			<CardHeader>
				<CardTitle>Account</CardTitle>
				<CardDescription>
					Make changes to your account here. Click submit when you are done.
				</CardDescription>
			</CardHeader>
			<CardContent className="space-y-2">
				<Form {...form}>
					<form onSubmit={form.handleSubmit(onSubmit)} className="space-y-8">
						<FormField
							control={form.control}
							name="name"
							render={({ field }) => (
								<FormItem>
									<FormLabel>Name</FormLabel>
									<FormControl>
										<Input placeholder="name" {...field} />
									</FormControl>
									<FormDescription>
										This is your public display name.
									</FormDescription>
									<FormMessage />
								</FormItem>
							)}
						/>
						<FormField
							control={form.control}
							name="username"
							render={({ field }) => (
								<FormItem>
									<FormLabel>Username</FormLabel>
									<FormControl>
										<Input placeholder="username" {...field} />
									</FormControl>
									<FormDescription>
										This will be used for logging in.
									</FormDescription>
									<FormMessage />
								</FormItem>
							)}
						/>
						<Button type="submit">Submit</Button>
					</form>
				</Form>
			</CardContent>
		</Card>
	)
}
