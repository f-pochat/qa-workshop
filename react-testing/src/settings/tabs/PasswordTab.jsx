import React from 'react'
import { Button } from "@/components/ui/button"
import {
	Card,
	CardContent,
	CardDescription,
	CardFooter,
	CardHeader,
	CardTitle,
} from "@/components/ui/card"
import { Input } from "@/components/ui/input"
import { z } from 'zod'
import { useToast } from '@/components/ui/use-toast'
import { useForm } from 'react-hook-form'
import { zodResolver } from '@hookform/resolvers/zod'
import { Form, FormControl, FormField, FormItem, FormLabel, FormMessage } from '@/components/ui/form'


const formSchema = z.object({
	currentPassword: z.string(),
	newPassword: z.string(),
	repeatPassword: z.string()
}).refine(data => data.newPassword !== data.currentPassword, {
	message: "New password must be different from current password",
	path: ["newPassword"]
}).refine(data => data.newPassword === data.repeatPassword, {
	message: "Passwords must match",
	path: ["repeatPassword"]
})

export const PasswordTab = () => {
	const { toast } = useToast();
	const form = useForm({
		resolver: zodResolver(formSchema),
		defaultValues: {
			currentPassword: "",
			newPassword: "",
			repeatPassword: ""
		}
	});

	const onSubmit = () => {
		toast({ title: "Success", text: "Password changed" })
	}

	return (
		<Card>
			<CardHeader>
				<CardTitle>Password</CardTitle>
				<CardDescription>
					Change your password here. After saving, you'll be logged out.
				</CardDescription>
			</CardHeader>
			<CardContent className="space-y-2">
				<Form {...form}>
					<form onSubmit={form.handleSubmit(onSubmit)} className="space-y-8">
						<FormField
							control={form.control}
							name="currentPassword"
							render={({ field }) => (
								<FormItem>
									<FormLabel>Current Password</FormLabel>
									<FormControl>
										<Input type="password" placeholder="Current Password" {...field} />
									</FormControl>
									<FormMessage />
								</FormItem>
							)}
						/>
						<FormField
							control={form.control}
							name="newPassword"
							render={({ field }) => (
								<FormItem>
									<FormLabel>New Password</FormLabel>
									<FormControl>
										<Input type="password" placeholder="New Password" {...field} />
									</FormControl>
									<FormMessage />
								</FormItem>
							)}
						/>
						<FormField
							control={form.control}
							name="repeatPassword"
							render={({ field }) => (
								<FormItem>
									<FormLabel>Repeat Password</FormLabel>
									<FormControl>
										<Input type="password" placeholder="Repeat Password" {...field} />
									</FormControl>
									<FormMessage />
								</FormItem>
							)}
						/>
						<Button type="submit">Submit</Button>
					</form>
				</Form>
			</CardContent>
		</Card>
	)
}
