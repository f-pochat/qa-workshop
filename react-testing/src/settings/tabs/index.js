export * from './AccountTab'
export * from './PasswordTab'