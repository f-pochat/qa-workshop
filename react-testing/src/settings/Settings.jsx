import React from 'react'
import {
  Tabs,
  TabsContent,
  TabsList,
  TabsTrigger,
} from '@/components/ui/tabs'
import { AccountTab, PasswordTab } from './tabs'
 
export const Settings = () => {
  return (
    <Tabs defaultValue="account" className="w-[400px]">
      <TabsList className="grid w-full grid-cols-2">
        <TabsTrigger value="account">Account</TabsTrigger>
        <TabsTrigger value="password">Password</TabsTrigger>
      </TabsList>
      <TabsContent value="account">
        <AccountTab />
      </TabsContent>
      <TabsContent value="password">
				<PasswordTab />
      </TabsContent>
    </Tabs>
  )
}