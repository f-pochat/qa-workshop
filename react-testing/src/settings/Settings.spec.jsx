import React from 'react'
import { describe, expect, test, vi, afterEach } from 'vitest'
import { render } from '@/test'
import { Settings } from './Settings'
import { AccountTab, PasswordTab } from './tabs'

// Shallow rendering child components.
// It's ideal to use if, say, AccountTab or PasswordTab have external dependencies
// such as a useDispatch redux hook which you don't care about when testing this parent component.
// 1) Pros: makes parent component snapshot smaller and therefore more readable.
// 2) Trade-off / Cons: you should have tests for AccountTab and PasswordTab to make sure they work as expected.
vi.mock('./tabs', () => ({
	AccountTab: vi.fn(() => 'AccountTab'),
	PasswordTab: vi.fn(() => 'PasswordTab')
}))

describe('Settings', () => {
	afterEach(() => {
		vi.clearAllMocks()
	})

	test('should match snapshot', () => {
		const { container } = render(<Settings />)
		expect(container.firstChild).toMatchSnapshot()
	})

	test('should call AccountTab and pass props', () => {
		render(<Settings />)
		expect(AccountTab).toHaveBeenCalledTimes(1)
		expect(AccountTab).toHaveBeenCalledWith({}, expect.anything())
	})

	// test.todo('should call PasswordTab and pass props', () => {})
})