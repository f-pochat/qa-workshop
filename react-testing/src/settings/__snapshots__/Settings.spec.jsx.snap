// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Settings > should match snapshot 1`] = `
<div
  class="w-[400px]"
  data-orientation="horizontal"
  dir="ltr"
>
  <div
    aria-orientation="horizontal"
    class="h-10 items-center justify-center rounded-md bg-muted p-1 text-muted-foreground grid w-full grid-cols-2"
    data-orientation="horizontal"
    role="tablist"
    style="outline: none;"
    tabindex="0"
  >
    <button
      aria-controls="radix-:r0:-content-account"
      aria-selected="true"
      class="inline-flex items-center justify-center whitespace-nowrap rounded-sm px-3 py-1.5 text-sm font-medium ring-offset-background transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow-sm"
      data-orientation="horizontal"
      data-radix-collection-item=""
      data-state="active"
      id="radix-:r0:-trigger-account"
      role="tab"
      tabindex="-1"
      type="button"
    >
      Account
    </button>
    <button
      aria-controls="radix-:r0:-content-password"
      aria-selected="false"
      class="inline-flex items-center justify-center whitespace-nowrap rounded-sm px-3 py-1.5 text-sm font-medium ring-offset-background transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow-sm"
      data-orientation="horizontal"
      data-radix-collection-item=""
      data-state="inactive"
      id="radix-:r0:-trigger-password"
      role="tab"
      tabindex="-1"
      type="button"
    >
      Password
    </button>
  </div>
  <div
    aria-labelledby="radix-:r0:-trigger-account"
    class="mt-2 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2"
    data-orientation="horizontal"
    data-state="active"
    id="radix-:r0:-content-account"
    role="tabpanel"
    style="animation-duration: 0s;"
    tabindex="0"
  >
    AccountTab
  </div>
  <div
    aria-labelledby="radix-:r0:-trigger-password"
    class="mt-2 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2"
    data-orientation="horizontal"
    data-state="inactive"
    hidden=""
    id="radix-:r0:-content-password"
    role="tabpanel"
    tabindex="0"
  />
</div>
`;
